linkerd/app/outbound/src/http/concrete.rs

```diff
@@ -23,8 +23,14 @@ use tracing::info_span;
 pub enum Dispatch {
     Balance(NameAddr, balance::EwmaConfig),
     Forward(Remote<ServerAddr>, Metadata),
+    Null,
 }
 
+/// A backend dispatcher was null.
+#[derive(Debug, thiserror::Error)]
+#[error("no backend dispatcher configured")]
```

Comment:
Is this error sufficiently informative when it shows up in logs and users ask about it? "backend dispatchers" are a bit of an internal concept.

Can this explanation be provided by the proxy alone? Or should the control plane provide this explanation?

Reply:
> Would this be more clearly expressed by something like `Dispatch::Fail { message: String }`? or even a type that holds an error type that is capable of configuring response codes, etc?

I also didn't like how the approach in this PR results in two different error paths, only one of which is descriptive, and also considered something like what you're describing. The reasoning behind the current approach is that the controller is currently communicating the desired error to the proxy using a `failureInjector` filter on the backend. The failure injector filter should inject the nicer error  before we actually hit the explicitly-failing backend stack, so the more descriptive error _is_ the one observed by the user, but the presence of the second error path definitely does feel unfortunate.

If we wanted to have an explicitly failing backend that fails with the error configured by the control plane (rather than an explicitly failing backend with a generic error message that _also_ has a `failureInjector` filter), it seems like there's one of two approaches:
1.  Special-case empty backends with `failureInjector` filters to return a `Dispatch::Fail` variant with the message and status code from the failure injector filter. This would avoid the use of the null backend with a generic error message, but it felt a little weird to synthesize something in the proxy that wasn't quite the same as the policy from the controller.
2. Introduce a `Dispatch::Fail`-like variant to the proxy API as well, and have the policy controller send that in this case, rather than a backend with a `None` dispatcher and a `failureInjector` filter. That would, in many ways, be the cleanest solution (IMO) but also adds more stuff to the proxy API to represent a configuration that _can_ be represented (although maybe less cleanly) with the existing API.

I went with the approach in this branch because it felt like the least intrusive solution, but would be happy to try one of the other approaches I discussed here if you think it would be better! I don't love this approach either...

Reply:
But the null backend fails with an error, right? Are these approaches really mutually exclusive?

My main argument is: I think it's surprising for something called a null backend to poll ready at all. If this backend is really unreachable, let's call it Unreachable and make it panic; or if it's going to return some sort of error, let's make that behavior explicit and part of how the backend is actually configured.

I don't much care whether filters or used or not. I just don't really like the semantics around the null backend.

Reply:
Oh, okay, got it.